server/src/main/java/org/elasticsearch/readiness/ReadinessService.java

```diff
@@ -277,7 +279,20 @@ private boolean isMasterElected(ClusterState clusterState) {
     // protected to allow mock service to override
     protected boolean areFileSettingsApplied(ClusterState clusterState) {
         ReservedStateMetadata fileSettingsMetadata = clusterState.metadata().reservedStateMetadata().get(FileSettingsService.NAMESPACE);
-        return fileSettingsMetadata != null && fileSettingsMetadata.version().equals(ReservedStateMetadata.NO_VERSION) == false;
+        if (fileSettingsMetadata == null) {
+            // In order to block readiness on file settings being applied, we need to know that the master node has written an initial
+            // version, or a marker that file settings don't exist. When upgrading from a version that did not have file settings, the
+            // current master node may not be the first node upgraded. To be safe, we wait to consider file settings application for
+            // readiness until the whole cluster supports file settings.
+            return supportsFileSettings(clusterState) == false;
+        } else {
+            return fileSettingsMetadata.version().equals(ReservedStateMetadata.NO_VERSION) == false;
+        }
+    }
+
+    @SuppressForbidden(reason = "need to check file settings support on exact cluster state")
+    private static boolean supportsFileSettings(ClusterState clusterState) {
+        return clusterState.clusterFeatures().clusterHasFeature(FileSettingsFeatures.FILE_SETTINGS_SUPPORTED);
```

Comment:
This is what `FeatureService.clusterHasFeature` does, no?

Reply:
Historical features are added by the FeatureService, so we can't check them here against a specific cluster state